Modules/vc-module-orders/VirtoCommerce.OrdersModule.Data/Services/DefaultCustomerOrderTotalsCalculator.cs

```diff
@@ -84,18 +93,24 @@ public virtual void CalculateTotals(CustomerOrder order)
 
             if (!order.InPayments.IsNullOrEmpty())
             {
-                order.PaymentTotal = order.InPayments.Sum(x => x.Total);
-                order.PaymentTotalWithTax = order.InPayments.Sum(x => x.TotalWithTax);
-                order.PaymentSubTotal = order.InPayments.Sum(x => x.Price);
-                order.PaymentSubTotalWithTax = order.InPayments.Sum(x => x.PriceWithTax);
-                order.PaymentDiscountTotal = order.InPayments.Sum(x => x.DiscountAmount);
-                order.PaymentDiscountTotalWithTax = order.InPayments.Sum(x => x.DiscountAmountWithTax);
-                order.DiscountTotal += order.InPayments.Sum(x => x.DiscountAmount);
-                order.DiscountTotalWithTax += order.InPayments.Sum(x => x.DiscountAmountWithTax);
-                order.TaxTotal += order.InPayments.Sum(x => x.TaxTotal);
+                order.PaymentTotal = order.InPayments.Any(x => x.Total.HasValue) ? order.InPayments.Sum(x => x.Total) : null;
+                order.PaymentTotalWithTax = order.InPayments.Any(x => x.TotalWithTax.HasValue) ? order.InPayments.Sum(x => x.TotalWithTax) : null;
+                order.PaymentSubTotal = order.InPayments.Any(x => x.Price.HasValue) ? order.InPayments.Sum(x => x.Price) : null;
+                order.PaymentSubTotalWithTax = order.InPayments.Any(x => x.PriceWithTax.HasValue) ? order.InPayments.Sum(x => x.PriceWithTax) : null;
+                order.PaymentDiscountTotal = order.InPayments.Any(x => x.DiscountAmount.HasValue) ? order.InPayments.Sum(x => x.DiscountAmount) : null;
+
+                var discountAmountWithTax = order.InPayments.Any(x => x.DiscountAmountWithTax.HasValue) ? order.InPayments.Sum(x => x.DiscountAmountWithTax) : null;
+                order.PaymentDiscountTotalWithTax = discountAmountWithTax;
+                order.DiscountTotalWithTax = discountAmountWithTax != null ? Convert.ToDecimal(order.DiscountTotalWithTax) + discountAmountWithTax : order.DiscountTotalWithTax;
+
+                var discountAmount = order.InPayments.Any(x => x.DiscountAmount.HasValue) ? order.InPayments.Sum(x => x.DiscountAmount) : null;
```

Comment:
Here I do not like addition of variable `discountAmount ` for calculation sharing. It decreases readability in our case and could lead to errors, and I suppose such optimization could be done automatically by compiler (not sure, but it is possible).